### ISSUE TITLE
canonicalization result incorrect if default `xmlns` presents and text node contains `&amp;`

### ISSUE THREAD

**Issue by colourful-land** [PeculiarVentures/xmldsigjs#47]
This javascript can reproduce the problem. Put it under the project root directory. It first try to canonicalize a very short XML piece with `xmlns`.

Run it and observe that `&amp;` in a text node is incorrectly replaced with `&amp;amp;`, then it shows by removing `xmlns` on the root element the canonicalization is correct.

The test code:

````
"use strict";

const XmlDSigJs = require("xmldsigjs");
var WebCrypto = require("node-webcrypto-ossl");
var crypto = new WebCrypto.Crypto();

XmlDSigJs.Application.setEngine("WebCrypto", crypto);

let xml1 = `<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en"><body><script type="text/javascript">
let message = " &amp; ETH";</script></body></html>`;

let xml2 = `<html xml:lang="en"><body><script type="text/javascript">
let message = " &amp; ETH";</script></body></html> `;


function c14n(xml) {
	const elem = new DOMParser().parseFromString(xml, "application/xml");
	const xmlCanonicalizer = new XmlDSigJs.XmlCanonicalizer(true, false); // comment / exclusive
	const res = xmlCanonicalizer.Canonicalize(elem);
	return res;
}

console.log(c14n(xml1))
console.log(c14n(xml2))

````

The output (observe the `&amp;amp;` part right before `ETH`):

````
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en"><body><script type="text/javascript">
let message = " &amp;amp; ETH";</script></body></html>
<html xml:lang="en"><body><script type="text/javascript">
let message = " &amp; ETH";</script></body></html>
````

**Comment by colourful-land** [comment]
This is more mysterious! If I change the test XML into this:

````
<x:html x:xmlns="http://www.w3.org/1999/xhtml" xml:lang="en"><x:body><x:script type="text/javascript">
let message = " &amp; ETH";</x:script></x:body></x:html>
````

That is, take the XML that can reproduce this bug, which has `xmlns`, replace it with `x:xmlns` and prefix every element with `x:`, then the canonicalization result is correct!
````
<x:html xmlns:x="" x:xmlns="http://www.w3.org/1999/xhtml" xml:lang="en"><x:body><x:script type="text/javascript">
let message = " &amp; ETH";</x:script></x:body></x:html>
````

**Comment by colourful-land** [comment]
Is there any hope that this is resolved? I hope we could find a way to pay to solve the bug, as our projects depend on a working xmldsig JavaScript library and there exists none in the world - PeculiarVentures' is the closest to a working version.

**Comment by rmhrisk** [comment]
We can look at it but @microshine is tied up with other work at the moment. We will get back to your issue when that’s done.

**Comment by microshine** [comment]
Looks like the problem is in `xmldom` module
https://github.com/xmldom/xmldom/issues/22

**Comment by colourful-land** [comment]
> Looks like the problem is in `xmldom` module
> [xmldom/xmldom#22](https://github.com/xmldom/xmldom/issues/22)

You are right on this, but xmldom closed that bug without resolving it. I left a comment there hoping it can be reopened

**Comment by colourful-land** [comment]
I'm closing this for now since I expect this issue to disappear if https://github.com/xmldom/xmldom/issues/203 is solved.

**Comment by micwallace** [comment]
This fix is now in pre-release @xmldom/xmldom@^0.9.0-beta.6

https://github.com/xmldom/xmldom/issues/203
